File: src/main/java/com/zepben/evolve/services/network/tracing/PhaseTrace.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.evolve.services.network.tracing;

import com.zepben.annotations.EverythingIsNonnullByDefault;
import com.zepben.evolve.cim.iec61970.base.core.ConductingEquipment;
import com.zepben.evolve.cim.iec61970.base.core.Terminal;
import com.zepben.evolve.cim.iec61970.base.wires.SinglePhaseKind;
import com.zepben.evolve.services.network.NetworkService;
import com.zepben.evolve.services.network.model.PhaseDirection;
import com.zepben.evolve.services.network.tracing.traversals.BasicTraversal;
import com.zepben.evolve.services.network.tracing.traversals.WeightedPriorityQueue;

import java.util.*;
import java.util.stream.Collectors;

/**
 * A class that creates commonly used phase based traces. You can add custom step actions and stop conditions
 * to the returned traversal.
 */
@SuppressWarnings("WeakerAccess")
@EverythingIsNonnullByDefault
class PhaseTrace {

    /**
     * @return a traversal that traces along phases in all directions through all open points.
     */
    public static BasicTraversal<PhaseStep> newTrace() {
        return newTrace(OpenTest.IGNORE_OPEN);
    }

    /**
     * @return a traversal that traces along phases in all directions stopping at normally open points.
     */
    public static BasicTraversal<PhaseStep> newNormalTrace() {
        return newTrace(OpenTest.NORMALLY_OPEN);
    }

    /**
     * @return a traversal that traces along phases in all directions stopping at currently open points.
     */
    public static BasicTraversal<PhaseStep> newCurrentTrace() {
        return newTrace(OpenTest.CURRENTLY_OPEN);
    }

    /**
     * @return a traversal that traces along phases in a downstream direction stopping at normally open points.
     */
    public static BasicTraversal<PhaseStep> newNormalDownstreamTrace() {
        return newDownstreamTrace(OpenTest.NORMALLY_OPEN, PhaseSelector.NORMAL_PHASES);
    }

    /**
     * @return a traversal that traces along phases in a downstream direction stopping at currently open points.
     */
    public static BasicTraversal<PhaseStep> newCurrentDownstreamTrace() {
        return newDownstreamTrace(OpenTest.CURRENTLY_OPEN, PhaseSelector.CURRENT_PHASES);
    }

    /**
     * @return a traversal that traces along phases in a upstream direction stopping at normally open points.
     */
    public static BasicTraversal<PhaseStep> newNormalUpstreamTrace() {
        return newUpstreamTrace(OpenTest.NORMALLY_OPEN, PhaseSelector.NORMAL_PHASES);
    }

    /**
     * @return a traversal that traces along phases in a upstream direction stopping at currently open points.
     */
    public static BasicTraversal<PhaseStep> newCurrentUpstreamTrace() {
        return newUpstreamTrace(OpenTest.CURRENTLY_OPEN, PhaseSelector.CURRENT_PHASES);
    }

    private static BasicTraversal<PhaseStep> newTrace(OpenTest isOpenTest) {
        return new BasicTraversal<>(queueNext(isOpenTest), WeightedPriorityQueue.processQueue(ps -> ps.phases().size()), new PhaseStepTracker());
    }

    private static BasicTraversal<PhaseStep> newDownstreamTrace(OpenTest isOpenTest, PhaseSelector activePhases) {
        return new BasicTraversal<>(queueNextDownstream(isOpenTest, activePhases),
            WeightedPriorityQueue.processQueue(ps -> ps.phases().size()),
            new PhaseStepTracker());
    }

    private static BasicTraversal<PhaseStep> newUpstreamTrace(OpenTest isOpenTest, PhaseSelector activePhases) {
        return new BasicTraversal<>(queueNextUpstream(isOpenTest, activePhases),
            WeightedPriorityQueue.processQueue(ps -> ps.phases().size()),
            new PhaseStepTracker());
    }

    private static BasicTraversal.QueueNext<PhaseStep> queueNext(final OpenTest openTest) {
        return (phaseStep, traversal) -> {
            if (phaseStep == null)
                return;

            Set<SinglePhaseKind> outPhases = new HashSet<>();
            phaseStep.conductingEquipment().getTerminals().forEach(terminal -> {
                outPhases.clear();
                for (SinglePhaseKind phase : phaseStep.phases()) {
                    if (!openTest.isOpen(phaseStep.conductingEquipment(), phase)) {
                        outPhases.add(phase);
                    }
                }

                queueConnected(traversal, terminal, outPhases);
            });
        };
    }

    private static BasicTraversal.QueueNext<PhaseStep> queueNextDownstream(final OpenTest openTest,
                                                                           PhaseSelector activePhases) {
        return (phaseStep, traversal) -> {
            if (phaseStep == null)
                return;

            Set<SinglePhaseKind> outPhases = new HashSet<>();
            phaseStep.conductingEquipment().getTerminals().forEach(terminal -> {
                outPhases.clear();
                getPhasesWithDirection(openTest, activePhases, terminal, phaseStep.phases(), PhaseDirection.OUT, outPhases);

                queueConnected(traversal, terminal, outPhases);
            });
        };
    }

    private static void queueConnected(BasicTraversal<PhaseStep> traversal, Terminal terminal, Set<SinglePhaseKind> outPhases) {
        if (!outPhases.isEmpty()) {
            NetworkService.connectedTerminals(terminal, outPhases).forEach(
                cr -> tryQueue(traversal, cr, cr.getToNominalPhases())
            );
        }
    }

    private static BasicTraversal.QueueNext<PhaseStep> queueNextUpstream(final OpenTest openTest,
                                                                         PhaseSelector activePhases) {
        return (phaseStep, traversal) -> {
            if (phaseStep == null)
                return;

            Set<SinglePhaseKind> inPhases = new HashSet<>();
            phaseStep.conductingEquipment().getTerminals().forEach(terminal -> {
                inPhases.clear();
                getPhasesWithDirection(openTest, activePhases, terminal, phaseStep.phases(), PhaseDirection.IN, inPhases);

                if (!inPhases.isEmpty()) {
                    List<ConnectivityResult> inTerminals = NetworkService.connectedTerminals(terminal, inPhases);
                    inTerminals.forEach(cr -> {
                        // When going upstream, we only want to traverse to connected terminals that have an out direction
                        Set<SinglePhaseKind> outPhases = cr.getToNominalPhases()
                            .stream()
                            .filter(phase -> activePhases.status(cr.getToTerminal(), phase).direction().has(PhaseDirection.OUT))
                            .collect(Collectors.toSet());

                        if (!outPhases.isEmpty()) {
                            tryQueue(traversal, cr, outPhases);
                        }
                    });
                }
            });
        };
    }

    private static void tryQueue(BasicTraversal<PhaseStep> traversal, ConnectivityResult cr, Collection<SinglePhaseKind> outPhases) {
        ConductingEquipment to = cr.getTo();
        if (to != null)
            traversal.queue().add(PhaseStep.continueAt(to, outPhases, cr.getFrom()));
    }

    private static void getPhasesWithDirection(OpenTest openTest,
                                               PhaseSelector activePhases,
                                               Terminal terminal,
                                               Set<SinglePhaseKind> candidatePhases,
                                               PhaseDirection direction,
                                               Set<SinglePhaseKind> matchedPhases) {
        ConductingEquipment conductingEquipment = Objects.requireNonNull(terminal.getConductingEquipment());
        for (SinglePhaseKind phase : candidatePhases) {
            if (terminal.getPhases().singlePhases().contains(phase) && !openTest.isOpen(conductingEquipment, phase)) {
                if (activePhases.status(terminal, phase).direction().has(direction)) {
                    matchedPhases.add(phase);
                }
            }
        }
    }

    // Should not be able to instantiate this class.
    private PhaseTrace() {
    }

}

File: src/main/kotlin/com/zepben/evolve/services/network/tracing/phases/PhaseTrace.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.evolve.services.network.tracing.phases

import com.zepben.evolve.cim.iec61970.base.core.Terminal
import com.zepben.evolve.cim.iec61970.base.wires.SinglePhaseKind
import com.zepben.evolve.services.network.NetworkService.Companion.connectedTerminals
import com.zepben.evolve.services.network.tracing.OpenTest
import com.zepben.evolve.services.network.tracing.connectivity.ConnectivityResult
import com.zepben.evolve.services.network.tracing.feeder.DirectionSelector
import com.zepben.evolve.services.network.tracing.feeder.FeederDirection
import com.zepben.evolve.services.network.tracing.phases.PhaseStep.Companion.continueAt
import com.zepben.evolve.services.network.tracing.traversals.BasicTraversal
import com.zepben.evolve.services.network.tracing.traversals.WeightedPriorityQueue.Companion.processQueue

/**
 * A class that creates commonly used phase based traces. You can add custom step actions and stop conditions
 * to the returned traversal.
 */
object PhaseTrace {

    /**
     * @return a traversal that traces along phases in all directions through all open points.
     */
    fun newTrace(): BasicTraversal<PhaseStep> = newTrace(OpenTest.IGNORE_OPEN)

    /**
     * @return a traversal that traces along phases in all directions stopping at normally open points.
     */
    fun newNormalTrace(): BasicTraversal<PhaseStep> = newTrace(OpenTest.NORMALLY_OPEN)

    /**
     * @return a traversal that traces along phases in all directions stopping at currently open points.
     */
    fun newCurrentTrace(): BasicTraversal<PhaseStep> = newTrace(OpenTest.CURRENTLY_OPEN)

    /**
     * @return a traversal that traces along phases in a downstream direction stopping at normally open points.
     */
    fun newNormalDownstreamTrace(): BasicTraversal<PhaseStep> =
        newDownstreamTrace(OpenTest.NORMALLY_OPEN, DirectionSelector.NORMAL_DIRECTION)

    /**
     * @return a traversal that traces along phases in a downstream direction stopping at currently open points.
     */
    fun newCurrentDownstreamTrace(): BasicTraversal<PhaseStep> =
        newDownstreamTrace(OpenTest.CURRENTLY_OPEN, DirectionSelector.CURRENT_DIRECTION)

    /**
     * @return a traversal that traces along phases in an upstream direction stopping at normally open points.
     */
    fun newNormalUpstreamTrace(): BasicTraversal<PhaseStep> =
        newUpstreamTrace(OpenTest.NORMALLY_OPEN, DirectionSelector.NORMAL_DIRECTION)

    /**
     * @return a traversal that traces along phases in an upstream direction stopping at currently open points.
     */
    fun newCurrentUpstreamTrace(): BasicTraversal<PhaseStep> =
        newUpstreamTrace(OpenTest.CURRENTLY_OPEN, DirectionSelector.CURRENT_DIRECTION)

    private fun newTrace(isOpenTest: OpenTest): BasicTraversal<PhaseStep> =
        BasicTraversal(queueNext(isOpenTest), processQueue { it.phases.size }, PhaseStepTracker())

    private fun newDownstreamTrace(isOpenTest: OpenTest, activeDirection: DirectionSelector): BasicTraversal<PhaseStep> =
        BasicTraversal(queueNextDownstream(isOpenTest, activeDirection), processQueue { it.phases.size }, PhaseStepTracker())

    private fun newUpstreamTrace(isOpenTest: OpenTest, activeDirection: DirectionSelector): BasicTraversal<PhaseStep> =
        BasicTraversal(queueNextUpstream(isOpenTest, activeDirection), processQueue { it.phases.size }, PhaseStepTracker())

    private fun queueNext(openTest: OpenTest): BasicTraversal.QueueNext<PhaseStep> =
        BasicTraversal.QueueNext { phaseStep, traversal ->
            val downPhases = mutableSetOf<SinglePhaseKind>()

            phaseStep.conductingEquipment.terminals.forEach {
                downPhases.clear()
                for (phase in phaseStep.phases) {
                    if (!openTest.isOpen(phaseStep.conductingEquipment, phase)) {
                        downPhases.add(phase)
                    }
                }

                queueConnected(traversal, it, downPhases)
            }
        }

    private fun queueNextDownstream(openTest: OpenTest, activeDirection: DirectionSelector): BasicTraversal.QueueNext<PhaseStep> =
        BasicTraversal.QueueNext { phaseStep, traversal ->
            phaseStep.conductingEquipment.terminals.forEach {
                queueConnected(traversal, it, getPhasesWithDirection(openTest, activeDirection, it, phaseStep.phases, FeederDirection.DOWNSTREAM))
            }
        }

    private fun queueConnected(traversal: BasicTraversal<PhaseStep>, terminal: Terminal, downPhases: Set<SinglePhaseKind>) {
        if (downPhases.isNotEmpty()) {
            connectedTerminals(terminal, downPhases).forEach {
                tryQueue(traversal, it, it.toNominalPhases)
            }
        }
    }

    private fun queueNextUpstream(openTest: OpenTest, activeDirection: DirectionSelector): BasicTraversal.QueueNext<PhaseStep> =
        BasicTraversal.QueueNext { phaseStep, traversal ->

            phaseStep.conductingEquipment.terminals.forEach { terminal ->
                val upPhases = getPhasesWithDirection(openTest, activeDirection, terminal, phaseStep.phases, FeederDirection.UPSTREAM)
                if (upPhases.isNotEmpty()) {
                    connectedTerminals(terminal, upPhases).forEach { cr ->
                        // When going upstream, we only want to traverse to connected terminals that have a DOWNSTREAM direction
                        if (activeDirection.select(cr.toTerminal).value.has(FeederDirection.DOWNSTREAM))
                            tryQueue(traversal, cr, cr.toNominalPhases)
                    }
                }
            }
        }

    private fun tryQueue(traversal: BasicTraversal<PhaseStep>, cr: ConnectivityResult, downPhases: Collection<SinglePhaseKind>) {
        cr.to?.let { traversal.queue.add(continueAt(it, downPhases, cr.from)) }
    }

    private fun getPhasesWithDirection(
        openTest: OpenTest,
        activeDirection: DirectionSelector,
        terminal: Terminal,
        candidatePhases: Set<SinglePhaseKind>,
        direction: FeederDirection
    ): Set<SinglePhaseKind> {
        val matchedPhases = mutableSetOf<SinglePhaseKind>()

        if (!activeDirection.select(terminal).value.has(direction))
            return matchedPhases

        val conductingEquipment = terminal.conductingEquipment!!
        for (phase in candidatePhases) {
            if (terminal.phases.singlePhases.contains(phase) && !openTest.isOpen(conductingEquipment, phase))
                matchedPhases.add(phase)
        }

        return matchedPhases
    }

}
